bpo-45723: Add macro for disabling/enabling CC warnings (GH-29466)

erlend-aasland opened the pull request:
- Add `PY_CHECK_CC_WARNING` helper 
- Normalise variable name: `ac_cv_[enable|disable]_[flag_name]_warning`; for example `ac_cv_disable_unused_result_warning`
- Use `AS_IF` to check results and add flags

No new cache vars added, so there is no speedups; only simplified ac code.

Bonus: fixed a misspelling bug (see `ac_cv_enable_enable_strict_prototypes_warning`)

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-45723](https://bugs.python.org/issue45723) -->
https://bugs.python.org/issue45723
<!-- /issue-number -->

Commit:
Add macro for disabling/enabling CC warnings

tiran commented on configure.ac:
```diff
@@ -1640,30 +1640,33 @@ AC_SUBST(LDFLAGS_NODIST)
-      CFLAGS_NODIST="$CFLAGS_NODIST -Wextra"
-    fi
+    PY_CHECK_CC_WARNING([enable], [extra])
+    AS_IF([test $ac_cv_enable_extra_warning = yes],
```
You can replace the test with ``AS_VAR_IF()``

```suggestion
    AS_VAR_IF([ac_cv_enable_extra_warning], [yes],
```

erlend-aasland replied:
I'll update the rest

Commit:
Use AS_VAR_IF iso. AS_IF

Co-authored-by: Christian Heimes <christian@python.org>

Commit:
Use AS_VAR_IF iso. AS_IF

Commit:
Use AS_VAR_IF for strict aliasing flag

Commit:
Use AS_CASE for disable unused result warning

Commit:
Make it possible to supply custom msg

Commit:
Remove newline

tiran commented on configure.ac:
```diff
@@ -1862,10 +1758,8 @@ yes)
-    then
-      CFLAGS_NODIST="$CFLAGS_NODIST -Werror=implicit-function-declaration"
-    fi
+    AS_IF([test $ac_cv_enable_implicit_function_declaration_error = yes],
```
```suggestion
    AS_VAR_IF([ac_cv_enable_implicit_function_declaration_error], [yes],
```

tiran commented on configure.ac:
```diff
@@ -1882,10 +1776,8 @@ yes)
-    then
-      CFLAGS_NODIST="$CFLAGS_NODIST -fvisibility=hidden"
-    fi
+    AS_IF([test $ac_cv_enable_visibility = yes],
```
```suggestion
    AS_VAR_IF([ac_cv_enable_visibility], [yes],
```

tiran commented on configure.ac:
```diff
@@ -1640,30 +1640,33 @@ AC_SUBST(LDFLAGS_NODIST)
+  AC_MSG_CHECKING(m4_ifblank([$3], [if we can $1 $CC $2 warning], [$3]))
+  AS_VAR_PUSHDEF([py_var], [ac_cv_$1_]m4_normalize($2)[_warning])
+  AS_VAR_COPY([py_cc], [CC])
+  AS_VAR_SET([CC], ["$CC -W$2 -Werror"])
```
Is ``CC`` the correct variable or should you rather override ``CFLAGS`` or ``BASECFLAGS``?

erlend-aasland replied:
The old code did override `CC`, but you are correct, it is cleaner to use `CFLAGS`. I'll fix it.

Commit:
Address review

- Use AS_VAR_IF() a couple of more times
- Use CFLAGS iso. CC when overriding compiler flags

tiran reviewed (approved):
LGTM, thanks!
